### DISCUSSION TITLE
Find best provider in Kademlia

### DISCUSSION THREAD

**Discussion by umgefahren** [libp2p/rust-libp2p#2185] [Q&A] [answered]
In my understanding of Kadmlia the location of a value to a key is determined by the nearest distance between a keys hash and a nodes id. Following that it must be possible to determine a node for providing a key, from a key. Is there a way to perform such a lookup?

**Comment by mxinden** [comment, marked as answer]
>  by the nearest distance between a keys hash and a nodes id.

 by the nearest distance between a keys hash and a nodes id **hash**.

> Following that it must be possible to determine a node for providing a key, from a key. Is there a way to perform such a lookup?

Do I understand correctly that you want the closest nodes only? In that case you can just use `Kademlia::find_closest`.

**Reply by umgefahren** [reply]
More specifically I want the node in a network which's node id hash is as closest to the keys hash.

**Reply by mxinden** [reply]
:+1: then `Kademlia::find_closest` should be the way to go. Let us know how it goes.
